File: server/services/auth_service.py
"""
Authentication service with JWT tokens
"""
from datetime import datetime, timedelta
from typing import Optional
from jose import JWTError, jwt
from passlib.context import CryptContext
from config import settings

# Password hashing
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

# In-memory user store (for demo - Firebase integration available)
users_db: dict = {}


def hash_password(password: str) -> str:
    """Hash a password"""
    return pwd_context.hash(password)


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a password against its hash"""
    return pwd_context.verify(plain_password, hashed_password)


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    """Create a JWT access token"""
    to_encode = data.copy()
    expire = datetime.utcnow() + (expires_delta or timedelta(hours=settings.JWT_EXPIRATION_HOURS))
    to_encode.update({"exp": expire})
    return jwt.encode(to_encode, settings.JWT_SECRET_KEY, algorithm=settings.JWT_ALGORITHM)


def decode_token(token: str) -> Optional[dict]:
    """Decode and verify a JWT token"""
    try:
        payload = jwt.decode(token, settings.JWT_SECRET_KEY, algorithms=[settings.JWT_ALGORITHM])
        return payload
    except JWTError:
        return None


def register_user(name: str, email: str, password: str, role: str = "student") -> dict:
    """Register a new user"""
    if email in users_db:
        raise ValueError("Email already registered")
    
    user_id = f"user_{len(users_db) + 1}"
    users_db[email] = {
        "id": user_id,
        "name": name,
        "email": email,
        "password": hash_password(password),
        "role": role,
        "created_at": datetime.utcnow().isoformat()
    }
    
    return {
        "id": user_id,
        "name": name,
        "email": email,
        "role": role
    }


def authenticate_user(email: str, password: str) -> Optional[dict]:
    """Authenticate a user and return user data if valid"""
    user = users_db.get(email)
    if not user:
        return None
    if not verify_password(password, user["password"]):
        return None
    return {
        "id": user["id"],
        "name": user["name"],
        "email": user["email"],
        "role": user["role"]
    }


def get_user_by_email(email: str) -> Optional[dict]:
    """Get user by email"""
    user = users_db.get(email)
    if user:
        return {
            "id": user["id"],
            "name": user["name"],
            "email": user["email"],
            "role": user["role"]
        }
    return None

File: server/src/services/authService.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { config } from '../config/env';
import { User, UserWithPassword } from '../types';

// In-memory user store (for demo)
const usersDb: Map<string, UserWithPassword> = new Map();

export const authService = {
    // Hash password
    async hashPassword(password: string): Promise<string> {
        return bcrypt.hash(password, 10);
    },

    // Verify password
    async verifyPassword(plainPassword: string, hashedPassword: string): Promise<boolean> {
        return bcrypt.compare(plainPassword, hashedPassword);
    },

    // Create JWT token
    createAccessToken(data: Record<string, any>): string {
        const expiresIn = `${config.jwtExpirationHours}h`;
        return jwt.sign(data, config.jwtSecret, {
            algorithm: config.jwtAlgorithm as jwt.Algorithm,
            expiresIn,
        });
    },

    // Verify and decode JWT token
    verifyToken(token: string): any {
        try {
            return jwt.verify(token, config.jwtSecret);
        } catch (error) {
            return null;
        }
    },

    // Register new user
    async registerUser(
        name: string,
        email: string,
        password: string,
        role: 'student' | 'instructor' | 'admin' = 'student'
    ): Promise<User> {
        if (usersDb.has(email)) {
            throw new Error('Email already registered');
        }

        const userId = `user_${usersDb.size + 1}`;
        const hashedPassword = await this.hashPassword(password);

        const user: UserWithPassword = {
            id: userId,
            name,
            email,
            password: hashedPassword,
            role,
            created_at: new Date().toISOString(),
        };

        usersDb.set(email, user);

        // Return user without password
        const { password: _, ...userWithoutPassword } = user;
        return userWithoutPassword;
    },

    // Authenticate user
    async authenticateUser(email: string, password: string): Promise<User | null> {
        const user = usersDb.get(email);
        if (!user) {
            return null;
        }

        const isValid = await this.verifyPassword(password, user.password);
        if (!isValid) {
            return null;
        }

        const { password: _, ...userWithoutPassword } = user;
        return userWithoutPassword;
    },

    // Get user by email
    getUserByEmail(email: string): User | null {
        const user = usersDb.get(email);
        if (!user) {
            return null;
        }

        const { password: _, ...userWithoutPassword } = user;
        return userWithoutPassword;
    },
};
